pkg/scheduler/queue/tenant_queues.go

```diff
@@ -117,42 +117,93 @@ func newQueueBroker(maxUserQueueSize int, forgetDelay time.Duration) *queueBroke
+// are not expected to occur when re-enqueuing a previously dequeued request.
+func (qb *queueBroker) enqueueRequestFront(r requestToEnqueue) error {
+	queue, err := qb.getOrAddTenantQueue(r.tenantID, r.maxQueriers)
+	if err != nil {
+		return err
+	}
+
+	queue.PushFront(r.req)
+	return nil
+}
+
+// getOrAddTenantQueue returns existing or new queue for tenant.
+// maxQueriers is used to compute which queriers should handle requests for this tenant.
+// If maxQueriers is <= 0, all queriers can handle this tenant's requests.
 // If maxQueriers has changed since the last call, queriers for this are recomputed.
-func (qb *queueBroker) getOrAddTenantQueue(tenantID TenantID, maxQueriers int) *list.List {
-	tenant := qb.tenantQuerierAssignments.getOrAddTenant(tenantID, maxQueriers)
-	if tenant == nil {
-		return nil
+func (qb *queueBroker) getOrAddTenantQueue(tenantID TenantID, maxQueriers int) (*list.List, error) {
+	_, err := qb.tenantQuerierAssignments.getOrAddTenant(tenantID, maxQueriers)
+	if err != nil {
+		return nil, err
 	}
 	queue := qb.tenantQueues[tenantID]
 
 	if queue == nil {
-		queue = &userQueue{
+		queue = &tenantQueue{
 			requests: list.New(),
 		}
 		qb.tenantQueues[tenantID] = queue
 	}
 
-	return queue.requests
+	return queue.requests, nil
+}
+
+func (qb *queueBroker) dequeueRequestForQuerier(lastTenantIndex int, querierID QuerierID) (Request, TenantID, int, error) {
+	tenantID, tenantIndex, err := qb.tenantQuerierAssignments.getNextTenantIDForQuerier(lastTenantIndex, querierID)
+	if err != nil {
+		return nil, tenantID, tenantIndex, err
+	}
+
+	tenantQueue := qb.tenantQueues[tenantID]
+	if tenantQueue == nil {
+		return nil, tenantID, tenantIndex, nil
+	}
+
+	// queue will be nonempty as empty queues are deleted
+	queueElement := tenantQueue.requests.Front()
+	tenantQueue.requests.Remove(queueElement)
+
+	if tenantQueue.requests.Len() == 0 {
+		qb.deleteQueue(tenantID)
+	}
+	return queueElement.Value, tenantID, tenantIndex, nil
+
 }
 
-// Finds next queue for the querier. To support fair scheduling between users, client is expected
-// to pass last user index returned by this function as argument. If there was no previous
-// last user index, use -1.
+// Finds next queue for the querier. To support fair scheduling between tenants, client is expected
+// to pass last tenant index returned by this function as argument. If there was no previous
+// last tenant index, use -1.
 //
 // getNextQueueForQuerier returns an error if the querier has already notified this scheduler that it is shutting down.
-func (qb *queueBroker) getNextQueueForQuerier(lastUserIndex int, querierID QuerierID) (*list.List, TenantID, int, error) {
-	nextTenantID, nextTenantIndex, err := qb.tenantQuerierAssignments.getNextTenantIDForQuerier(lastUserIndex, querierID)
+func (qb *queueBroker) getNextQueueForQuerier(lastTenantIndex int, querierID QuerierID) (*list.List, TenantID, int, error) {
```

Comment:
This method seems to only be used in tests now - can it be removed?

Reply:
removed